module/zfs/bpobj.c

```diff
@@ -242,119 +279,167 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
+		mutex_enter(&bpo->bpo_lock);
+		ASSERT(bpobj_is_open(bpo));
+		if (free)
+			dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+
+		err = bpobj_iterate_blkptrs(bpi, func, arg, tx, free);
+		if (err != 0) {
+			mutex_exit(&bpo->bpo_lock);
+			/* push back onto stack for cleanup function */
+			list_insert_head(&stack, bpi);
+			goto out;
 		}
-
-		ASSERT3U(offset, >=, dbuf->db_offset);
-		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
-
-		objarray = dbuf->db_data;
-		err = bpobj_open(&sublist, bpo->bpo_os, objarray[blkoff]);
-		if (err)
-			break;
-		if (free) {
-			err = bpobj_space(&sublist,
-			    &used_before, &comp_before, &uncomp_before);
-			if (err != 0) {
-				bpobj_close(&sublist);
-				break;
+		/* A subobj with no subobjs below it can be cleaned up. */
+		if (bpo->bpo_phys->bpo_num_subobjs == 0 ||
+		    !bpo->bpo_havesubobj ||
+		    bpo->bpo_phys->bpo_subobjs == 0) {
+			/*
+			 * If there are no entries, there should
+			 * be no bytes.
+			 */
+			if (bpobj_is_empty(bpo)) {
+				ASSERT0(bpo->bpo_phys->bpo_bytes);
+				ASSERT0(bpo->bpo_phys->bpo_comp);
+				ASSERT0(bpo->bpo_phys->bpo_uncomp);
 			}
+			if (free && (bpi->bpi_parent != NULL)) {
+				bpobj_t *p = bpi->bpi_parent->bpi_bpo;
+
+				ASSERT0(bpo->bpo_phys->bpo_num_blkptrs);
+				ASSERT3U(p->bpo_phys->bpo_num_subobjs, >, 0);
+				ASSERT3U(bpi->bpi_index, ==,
+				    p->bpo_phys->bpo_num_subobjs - 1);
+
+				p->bpo_phys->bpo_num_subobjs--;
+
+				VERIFY3U(0, ==, dmu_free_range(p->bpo_os,
+				    p->bpo_phys->bpo_subobjs,
+				    bpi->bpi_index * sizeof (uint64_t),
+				    sizeof (uint64_t), tx));
+
+				/* eliminate the empty subobj */
+				if (bpo->bpo_havesubobj &&
+				    bpo->bpo_phys->bpo_subobjs != 0) {
+					ASSERT0(bpo->bpo_phys->bpo_num_subobjs);
+					err = dmu_object_free(bpo->bpo_os,
+					    bpo->bpo_phys->bpo_subobjs, tx);
+					if (err) {
+						mutex_exit(&bpo->bpo_lock);
+						break;
+					}
+					dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+					bpo->bpo_phys->bpo_subobjs = 0;
+				}
+				err = dmu_object_free(p->bpo_os,
+				    bpi->bpi_object, tx);
+				if (err) {
+					mutex_exit(&bpo->bpo_lock);
+					break;
+				}
+			}
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
+			}
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			continue;
 		}
-		err = bpobj_iterate_impl(&sublist, func, arg, tx, free);
-		if (free) {
-			VERIFY3U(0, ==, bpobj_space(&sublist,
-			    &used_after, &comp_after, &uncomp_after));
-			bpo->bpo_phys->bpo_bytes -= used_before - used_after;
-			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
-			bpo->bpo_phys->bpo_comp -= comp_before - comp_after;
-			bpo->bpo_phys->bpo_uncomp -=
-			    uncomp_before - uncomp_after;
-		}
+		ASSERT(bpo->bpo_havecomp);
+		/*
+		 * Can't clean up bpobj while it still has subobjs.  Place it
+		 * back in the list ahead of its children.  It will be cleaned
+		 * up once the children are all freed.
```

Comment:
The general structure of code inside the loop is:
```
bpobj_iterate_blkptrs()
...
if (no more work to do on this one) {
  ...
  continue;
}
/* still need to process subobj's */
...
```
I wonder if it would be more readable to structure this like:
```
bpobj_iterate_blkptrs()
...
if (no more work to do on this one) {
 ...
} else {
  /* still need to process subobj's */
  ...
}
```

Reply:
Ok.  More indenting but it's better than having to spot the continue statement when just reading the code.